src/main/java/org/commcare/session/RemoteQuerySessionManager.java

```diff
@@ -117,33 +117,56 @@ public URL getBaseUrl() {
     }
 
     /**
-     * @param skipDefaultPromptValues don't apply the default value expressions for query prompts
+     * Evaluate filters to be applied to the search uri using following rules -
+     * 1. If no defaults are specified for a property, add the user input for search property as it is
+     * 2. If defaults are specified, only add the user input if it's also specified in defaults
+     * 3. If no user input is specified, add the defaults for the property as it is
+     *
+     * @param skipUserInput don't populate user inputs for query prompts into search params
      * @return filters to be applied to case search uri as query params
      */
-    public Multimap<String, String> getRawQueryParams(boolean skipDefaultPromptValues) {
-        Multimap<String, String> params = ArrayListMultimap.create();
+    public Multimap<String, String> getRawQueryParams(boolean skipUserInput) {
+        Multimap<String, String> defaultParams = ArrayListMultimap.create();
         Multimap<String, XPathExpression> hiddenQueryValues = queryDatum.getHiddenQueryValues();
         for (String key : hiddenQueryValues.keySet()) {
             for (XPathExpression xpathExpression : hiddenQueryValues.get(key)) {
                 String evaluatedExpr = evalXpathExpression(xpathExpression, evaluationContext);
-                params.put(key, evaluatedExpr);
+                defaultParams.put(key, evaluatedExpr);
             }
         }
 
-        if (!skipDefaultPromptValues) {
+        Multimap<String, String> searchParams = ArrayListMultimap.create();
+
+        // Populate User Params
+        if (!skipUserInput) {
             for (Enumeration e = userAnswers.keys(); e.hasMoreElements(); ) {
                 String key = (String)e.nextElement();
                 String value = userAnswers.get(key);
-                if (!(params.containsKey(key) && params.get(key).contains(value))) {
-                    if (value != null) {
-                        params.put(key, userAnswers.get(key));
+
+                // if defaults are defined, make sure userParams are restricted to defaults
+                String[] choices = RemoteQuerySessionManager.extractMultipleChoices(value);
+                for (String choice : choices) {
+                    if (choice != null &&
+                            ((defaultParams.containsKey(key) && defaultParams.get(key).contains(choice))
+                                    || !defaultParams.containsKey(key)) &&
+                            !(searchParams.containsKey(key) && searchParams.get(key).contains(choice))) {
```

Comment:
hard to follow the logic here - would be useful to break it up into some separate vars:
```java
if (choice == null) {
    continue
}

boolean noDefault = !defaultParams.containsKey(key)
boolean validDefault = defaultParams.containsKey(key) && defaultParams.get(key).contains(choice)
boolean duplicateValue = searchParams.containsKey(key) && searchParams.get(key).contains(choice)

if ( (noDefault || validDefault) && !duplicateValue) {
    searchParams.put(key, choice);
}
```

Incidentally `noDefault || validDefault` (`!defaultParams.containsKey(key) || (defaultParams.containsKey(key) && defaultParams.get(key).contains(choice))`) can be simplified to `!defaultParams.containsKey(key) || defaultParams.get(key).contains(choice)` so we can make this:

```java
boolean validParam = !defaultParams.containsKey(key) || defaultParams.get(key).contains(choice)
boolean duplicateValue = searchParams.containsKey(key) && searchParams.get(key).contains(choice)

if (validParam && !duplicateValue) {
    searchParams.put(key, choice);
}
```

Explanation: `!A or (A & B)` -> `(!A or A) & (!A or B)` -> `True & (!A or B)` -> `!A or B`

Reply:
thanks this is so much better. https://github.com/dimagi/commcare-core/pull/1035/commits/6d82f12ce3803181464b6dca0c7340ecca73f645